File: Pfe_Backend/src/main/java/com/example/demo/entities/Comment.java
package com.example.demo.entities;

import lombok.*;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Document(collection = "comments")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Comment {

    @Id
    private String commentId;

    private String courseId;

    private String authorId;        // userId of commenter
    private String authorUsername;
    private String authorRole;      // "STUDENT", "INSTRUCTOR", "ADMIN"
    private String authorPhoto;     // photo file id when available

    private String content;

    @Builder.Default
    private LocalDateTime createdAt = LocalDateTime.now();
    private LocalDateTime updatedAt;

    /** null = top-level comment; non-null = reply to another comment */
    private String parentCommentId;

    /** userIds of users who liked this comment */
    @Builder.Default
    private List<String> likedByUserIds = new ArrayList<>();
}

File: mobile_front/lib/models/comment.dart
class Comment {
  final String commentId;
  final String courseId;
  final String authorId;
  final String authorUsername;
  final String? authorRole;
  final String? authorPhoto;
  final String content;
  final String? createdAt;
  final String? parentCommentId;
  final List<String> likedByUserIds;

  Comment({
    required this.commentId,
    required this.courseId,
    required this.authorId,
    required this.authorUsername,
    this.authorRole,
    this.authorPhoto,
    required this.content,
    this.createdAt,
    this.parentCommentId,
    this.likedByUserIds = const [],
  });

  factory Comment.fromJson(Map<String, dynamic> json) {
    return Comment(
      commentId: json['commentId'] ?? '',
      courseId: json['courseId'] ?? '',
      authorId: json['authorId'] ?? '',
      authorUsername: json['authorUsername'] ?? 'Unknown',
      authorRole: json['authorRole'],
      authorPhoto: json['authorPhoto'],
      content: json['content'] ?? '',
      createdAt: json['createdAt']?.toString(),
      parentCommentId: json['parentCommentId'],
      likedByUserIds: json['likedByUserIds'] != null
          ? List<String>.from(json['likedByUserIds'])
          : [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'commentId': commentId,
      'courseId': courseId,
      'authorId': authorId,
      'authorUsername': authorUsername,
      'authorRole': authorRole,
      'authorPhoto': authorPhoto,
      'content': content,
      'createdAt': createdAt,
      'parentCommentId': parentCommentId,
      'likedByUserIds': likedByUserIds,
    };
  }
}
